enhancements/ingress/custom-route-configuration.md

```diff
@@ -0,0 +1,242 @@
+
+### Non-Goals
+
+1. provide a way to specify SNI serving certificates for operands.
+2. dictate how an operator must expose itself, the termination policy on a route, or how the secret is used.
+
+## Proposal
+
+```
+type IngressSpec struct {
+    // other fields
+
+    // CustomizeableRoutes is a list of routes that a cluster-admin wants to customize.  It is logically keyed by
+    // .spec.customizeableRoutes[index].{namespace,name}.
+    // To determine the set of possible keys, look at .status.customizeableRoutes where participating operators place
+    // current route status keyed the same way.
+    CustomizeableRoutes []CustomizeableRouteSpec
+}
+
+type IngressStatus struct {
+    // other fields
+
+    // CustomizeableRoutes is where participating operators place the current route status for routes which the cluster-admin
+    // can customize hostnames and serving certificates.
+    // How the operator uses that serving certificate is up to the individual operator.
+    CustomizeableRoutes []CustomizeableRouteStatus
+}
+
+type CustomizeableRouteSpec struct{
+    // namespace is the namespace of the route to customize.  It must be a real namespace.
+    Namespace string
+    // name is the *logical* name of the route to customize.  It does not have to be the actual name of a route resource.
+    // Keep in mind that this is your API for users to customize.  You could later rename the route, but you cannot rename
+    // this name.
+    Name string
+    // desiredHost is the name that a cluster-admin wants to specify
+    DesiredHost string
+    // ServingCertKeyPairSecret is a reference to a secret in namespace/openshift-config that is a kubernetes tls secret.
+    // The serving cert/key pair must match and will be used by the operator to fulfill the intent of serving with this name.
+    // That means it could be embedded into the route or used by the operand directly.
+    // Operands should take care to ensure that if they use passthrough termination, they properly use SNI to allow service
+    // DNS access to continue to function correctly.
+    ServingCertKeyPairSecret SecretNameReference
+
+    // possible future, we could add a set of SNI mappings.  I suspect most operators would not properly handle it today.
+}
+
+type CustomizeableRouteStatus struct{
+    // namespace is the namespace of the route to customize.  It must be a real namespace.
+    Namespace string
+    // name is the *logical* name of the route to customize.  It does not have to be the actual name of a route resource.
+    // Keep in mind that this is your API for users to customize.  You could later rename the route, but you cannot rename
+    // this name.
+    Name string
+    // defaultHost is the normal name of this route.
+    DefaultHost string
+    // ConsumingUsers is a slice of users that need to have read permission on the secrets in order to use them.
+    // This will usually be an operator service account.
+    ConsumingUsers []string
```

Comment:
Why is this in status? Can this be configured? How?

Reply:
I suppose the ingress-operator will be setting those statuses ?

Reply:
The section "Control loop to manage secret read permissions in openshift-config" references this field.  As I understand it, the operator that manages the route and its associated secrets sets this field to specify that its or its operand's service account needs access to the secret so that the operator can read the certificate and inject it into the route or into the operand.  For example, the console operator needs to read the secret in order to build the console route, and if the console route were passthrough, the console operator might also need inject the certificate into the console pod.  So the console operator needs to specify the console operator's service account in `ConsumingUsers` for the console CustomizeableRoute.